src/JuMP.jl

```diff
@@ -335,21 +335,104 @@ Base.broadcastable(model::Model) = Ref(model)
+model depends on which operating mode JuMP is in (see [`mode`](@ref)).
 
-If JuMP is in `DIRECT` mode (i.e., the model was created using
-[`direct_model`](@ref)), the backend will be the optimizer passed to
-[`direct_model`](@ref).
+ * If JuMP is in `DIRECT` mode (i.e., the model was created using
+   [`direct_model`](@ref)), the backend will be the optimizer passed to
+   [`direct_model`](@ref).
+ * If JuMP is in `MANUAL` or `AUTOMATIC` mode, the backend is a
+   `MOI.Utilities.CachingOptimizer`.
 
-If JuMP is in `MANUAL` or `AUTOMATIC` mode, the backend is a
-`MOI.Utilities.CachingOptimizer`.
+**This function should only be used by advanced users looking to access
+low-level MathOptInterface or solver-specific functionality.**
 
-This function should only be used by advanced users looking to access low-level
-MathOptInterface or solver-specific functionality.
+## Notes
+
+If JuMP is not in `DIRECT` mode, the type returned by `backend` may change
+between any JuMP releases. Therefore, only use the public API exposed by
+MathOptInterface, and do not access internal fields. If you require access to
+the innermost optimizer, see [`unsafe_backend`](@ref). Alternatively, use
+[`direct_model`](@ref) to create a JuMP model in `DIRECT` mode.
+
+See also: [`unsafe_backend`](@ref).
 """
 backend(model::Model) = model.moi_backend
 
+"""
+    unsafe_backend(model::Model)
+
+Return the innermost optimizer associated with the JuMP model `model`.
+
+**This function should only be used by advanced users looking to access
+low-level solver-specific functionality. It has a high-risk of incorrect usage.
+We strongly suggest you use the alternative suggested below.**
+
+See also: [`backend`](@ref).
+
+## Unsafe behavior
+
+This function is unsafe because the innermost optimizer may not represent the
+same problem as the JuMP `model`.
+
+Thus, before calling `unsafe_backend` you should first call
+[`MOI.Utilities.attach_optimizer`](@ref) to ensure that the backend is
+synchronized with the JuMP model.
+```julia
+MOI.Utilities.attach_optimizer(model)
+inner = unsafe_backend(model)
+```
+
+Moreover, if you modify the JuMP model, the reference you have to the backend
+(i.e., `inner` in the example above) may be out-dated. Always get a new
+reference by calling `unsafe_backend` before calling any low-level
+functionality.
+
+This function is also unsafe in the reverse direction: if you modify the unsafe
+backend, e.g., by adding a new constraint to `inner`, the changes may be
+silently discarded by JuMP when the JuMP `model` is modified or solved.
+
+## Alternative
+
+Instead of `unsafe_backend`, create a model using [`direct_model`](@ref) and
+call [`backend`](@ref) instead.
+
+For example, instead of:
+```julia
+model = Model(GLPK.Optimizer)
+@variable(model, x >= 0)
+MOI.Utilities.attach_optimizer(model)
+glpk = unsafe_backend(model)
+# ... use GLPK ...
+@objective(model, Min, 2x + 1)
+MOI.Utilities.attach_optimizer(model)
+glpk = unsafe_backend(model)
```

Comment:
Why does this need to be called again? In which case would this new `glpk` reference a different object?

Reply:
I guess I'm overthinking this. It's actually less unsafe than I thought. It's the inner pointer that is unsafe, not the `glpk` object.